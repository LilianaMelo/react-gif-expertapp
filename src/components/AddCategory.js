import React, { useState } from "react";
import { PropTypes } from "prop-types";


//GifExpertApp
// (props) es una forma de hacerlo
// se desestructura los argumentos del componenete
export const AddCategory = ( {setCategories} ) => {
    
    const [inputValue, setInputValue] = useState(""); // la caja de texto tiene un estado, para saber lo que la persona esta escribiendo (input :: Add Category). 
    // inputValue:: valor de entrada, setInputValue:: establecer valor de entrada.
    
    // inputValue:: por defecto tiene el valor de Hola Mundo, un string, un arreglo o un objeto.

    //const [gifRandom, setGifRandom] = useState("");
    
    const handleInputChange = (e) => {  // extraer la informacion de onChange y recibimos el evento (e).

        //console.log(e.target.value);
        setInputValue(e.target.value); // con esto permite hacer cambios en el input o barra.

        //console.log("handleInputChange llamado prueba")
    }

    const handleSubmit = (e) => {

        e.preventDefault(); // prevenir el comportamiento por defecto del navegador.
        //console.log("handleSubmit", inputValue);

        // .trim :: elimina los espacios en blanco al inicio y al final del string.

        if ( inputValue.trim().length > 2 ) {
            setCategories( categorias => [ inputValue, ...categorias ]); // los ... permiten agregar un nuevo item. 
            setInputValue("");
        } else {
            
            alert("No se encontro: " + inputValue +", intente nuevamente con otra palabra");
        } 
        
    }
    
    /*
    const changeSubmit = (e) => {

        e.preventDefault(); // prevenir el comportamiento por defecto del navegador.
        
        if (gifRandom) {
            setCategories(categorias => [gifRandom, categorias]);
            setGifRandom("");
        }    
    } */
    
    
    return ( // primero hace el return y muestra la informaci√≥n // snapshot
        <form onSubmit={handleSubmit}>  {/* el form y el onSubmit quita el refres de la pagina*/} {/* se llama a la constante*/}

            {/* <h1>{inputValue}</h1> esto muestra en una linea lo que se escribe en el input.*/}
            
            {/*<p>{inputValue}</p>  se borra esta opcion */}
            {/*<h2>Add Category</h2> se borra esta opcion  */}

            <input 
                type="text" // cada uno de estos son las props
                placeholder="Buscar gifs"
                value={inputValue} // muestra el valor de Hola Mundo, pero no deja modificar el input.
                onChange={ handleInputChange } // es para cambiar el valor o el contenido del input. // e es el evento // Luego llamamos a handleInputChange para 
                /* se llama a la constante*/
            />

            <div className="container"> 
                <button onSubmit={handleSubmit}>Buscar</button>
                <br />

                <button //  onSubmit={ changeSubmit }
                >
                    Gif Aleatorio
                </button>

            </div>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired // definicion de 
}



