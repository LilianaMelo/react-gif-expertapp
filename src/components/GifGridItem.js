import React from "react";
import { PropTypes } from "prop-types";

// export const GifGridItem = (props) => {
export const GifGridItem = ({ title, url}) => {// tambien se puede hacer asi es mÃ¡s entendible. => {
    //console.log(props);
    //console.log({id, title, url});

    return ( // aqui se deja el div para poner o retornar una clase.
        // className = es de estilos en React clase de css.

        <div className="card animate__animated animate__bounce">
            <img src={url} alt={title} />
            <p> {title} </p> 
        </div>
    ) // title es del gifs.
}


// agregar propytes aqui:
GifGridItem.propType = { // sale error en la consola de la prueba, indicando que title y url son requeridos u obligatorios.
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}
