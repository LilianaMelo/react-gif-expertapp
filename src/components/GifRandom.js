
export const GifRandom  = async (category) =>  {

    // trae las imgs aleatorias.
    
    const url = `https://api.giphy.com/v1/gifs/random?q=${encodeURI(category)}&rating=g&api_key=CgsE47b51eLpijMbALWycMxgZ76uzGiP&limit=1 `;

    const resp = await fetch(url);// busca la url.
    const {data} = await resp.json(); // info en formato json

    const randomImg = data.map( img => { // extraer informaciÃ³n.

        return {
            id: img.id,
            title: img.title,
            url: img.images?.downsized.url // images? es para preguntar que si vienen las imagenes que las utilice y las muestre.

        }
    }) 

    return randomImg;
    //console.log(randomImg);
    // .then((random_id) => console.log(random_id));

}
