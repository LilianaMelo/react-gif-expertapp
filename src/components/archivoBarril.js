// Archivo de barril: 
// es una forma de todo en uno. 
// cuando se necesiten todos los componentes se llama este archivo.

// en el video el nombre del archivo es index.js

// este archivo solo sirve en react 18

// se debe nombrar como index.js


export * from './AddCategory';
export * from './GifGrid';
export * from './GifGridItem';