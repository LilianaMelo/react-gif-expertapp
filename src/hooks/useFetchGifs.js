
//use significa que es un hook
// hook es una función. y tiene el trabajo más pesado.

import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";
//import { GifRandom } from "../components/GifRandom";


// El CustomHook es una funcion component y se extrae una pequeña logica para implementar una carga automatica.

export const useFetchGifs = (category) => { // se trae el category o los props dependiendo como se esta haciendo.

    // estado inicial
    const [state, setState] = useState ({ // estado, indicarle a otros componentes que lo utilicen cuando algo cambio.

        data: [],
        loading: true

    });

    // los efectos no pueden ser async por esperan algo asincrono.
    
    //aqui dispara un efecto
    useEffect( () => { // en los customhook tambien se pueden hacer efectos. useEffect se realiza cuando cambia la categoria.

        // Tarea asincrona  // petición http
        getGifs(category)// pasa una promesa
            .then( imgs => {  

                setState({ // cambia la información.
                    data: imgs,
                    loading: false
                }); 
            }) 

        // petición http
            /* GifRandom(category) // promesa se pasa al useFecthGifs. desde GifRandom.
            .then( imgRandom => {

                setState({
                    data: imgRandom,
                    loading: false
                });
            }) */

    }, [category]) // aqui se llama la category 


    /*
    setTimeout( () => {
        setState({ // setState es del arreglo de arriba de customHook
            data: [1,2,3,4,5,6,7,8,9,0],
            loading: false
        })
    }, 3000); // esto indica que va a ejecutar lo que esta adentro(setTimeout) a los 3 segundo s

    */
    
    return state; // {data: [], loading: true}  el state es la informacion que tiene el useState... ok
}
