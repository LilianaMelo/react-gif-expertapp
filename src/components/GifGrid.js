//import React, { useState, useEffect } from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
//import { getGifs } from "./helpers/getGifs";


// coleccion de todos los elementos de esa categoria. // componente que recibe la info de addCategory

export const GifGrid = ({category}) => {


    //CustomHook:

    const {data:images, loading} = useFetchGifs(category);


    
     /* const [images, setImages] = useState([]); // la coleccion por defecto es un arreglo[].

    // useEffect permite ejecutar cierto codigo de forma condicional 
    
    useEffect( () => { // aqui se dispara una sola vez.
        getGifs(category)// aquÃ­ solo se ejecuta getGifs() una sola vez.
            .then(setImages) ; 
    }, [category])  // un arreglo o listas de dependencias:: React.DependencyList */
    


    /*
    // peticion http 
    const getGifs = async () => { //  funciones

        // traer imagenes que corresponden a la categoria.
        // peticion de http en javascript
        const url = `https://api.giphy.com/v1/gifs/search?q=${ encodeURI( category ) }&limit=10&api_key=CgsE47b51eLpijMbALWycMxgZ76uzGiP`; 

        //  rick and morty se reemplaza por ${ encodeURI( category ) } para especificar la busqueda ej: dragon ball. //es una peticion.

        const respuesta = await fetch(url); // busca la url.
        const {data} = await respuesta.json();// informacion en formato json.

        const gifs = data.map( img => { // extraer informacion
            return{
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url // images? es para preguntar que si vienen las imagenes que las utilice y las muestre.
            }
        })

        console.log(gifs);
        setImages(gifs);
    }
    */

    //getGifs(); esto genera hace que se ejecute nuevamente toda la funcion de getGifs
    
    // contenedor de cartas: card-grid
    return (
        <>
            <h3 className="card animate__animated animate__fadeIn"> {category} </h3>

            {loading && <p className="animate__flash">Cargando</p>}

            <div className="card-grid">

                {
                    images.map( img => (
                        <GifGridItem 
                            key={img.id}
                            { ...img} 
                        />
                    ))
                }

                {/* images.map( img => (
                        <li key={img.id}> { img.title } </li>
                    )) esto para hacer la prueba antes de hacer el llamado del componente GifGridItem  * */}

            </div> 
        </>
    )
}