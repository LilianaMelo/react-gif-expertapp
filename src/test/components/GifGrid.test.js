import React from "react";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock('../../hooks/useFetchGifs'); // finguir cualquier llamado a ese archivo, y suponer y controlar la informacion que eso va a responder

// 1. mock son funciones simuladas, permiten probar los enlaces entre el código, borrando la implementación real de una función


// Pruebas unitarias.
describe('Pruebas en el <GifGrid />', () => {

    const category = "Dragon Ball";

    test('debe mostrarse correctamente', () => {

        // aqui se falsea la data, es decir no se esta usando el componente de  useFetchGifs, para este caso de la prueba.
        
        // falsa implementacion. 1.
        
        useFetchGifs.mockReturnValue({ // retorna el valor de la funcion simulada. (mock)

            //aqui va el objeto lo que devuelve la funcion useFetchGifs
            data: [],
            loading: true
        });  

        const wrapper = shallow(<GifGrid category={category}/>); // wranner nombre de la constante puede cambiar. 

        expect(wrapper).toMatchSnapshot();

    });


    test('debe mostrar items cuando se cargan imágenes useFetchGifs', () => { 

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/casa/verde.jpg',
            title: 'Cualquier cosa'
        }];

        // mock imitar, fingir algo, para que el componente aqui crea que tiene la informacion o la data con el custom hook cuando ya la cargo.

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });  


        const wrapper = shallow(<GifGrid  category={category} /> );

        //expect(wrapper).toMatchSnapshot(); //no se evalua con el MatchSnapshot().

        // busque todos los parrafos en html existentes. toBe que se espera (false)
        
        expect(wrapper.find('p').exists()).toBe(false);

        // busque los gifs y saber cuantos elementos hay. y esperaria la cantidad de gifs.
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
         // con estas dos expect son más precisas las pruebas.
    })

})