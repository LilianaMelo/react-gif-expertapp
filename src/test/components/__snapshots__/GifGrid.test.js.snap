// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid /> debe mostrarse correctamente 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
     
    Dragon Ball
     
  </h3>
  <p
    className="animate__flash"
  >
    Cargando
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
